qa: Fix TxIndex race conditions

hodlinator opened the pull request:
- Add synchronization in 3 places where if the Transaction Index happens to be slow, we get rare test failures when querying it for transactions (one such case experienced on Windows, prompting investigation).
- Remove unnecessary TxIndex initialization in some tests.
- Add some test coverage where TxIndex aspect could be tested in feature_init.py.

hodlinator committed:
qa: Remove unnecessary -txindex args

(Parent commit ensured indexes in feature_init.py are actually used, otherwise they would be removed here as well).

mzumsande commented on test/functional/feature_init.py:
```diff
@@ -86,12 +86,11 @@ def check_clean_start():
         if self.is_wallet_compiled():
             lines_to_terminate_after.append(b'Verifying wallet')
 
-        for terminate_line in lines_to_terminate_after:
```
why unnecessary? The point of the test (which has a very different "philosophy" from most of our other functional tests) is to trigger these restarts on purpose, simulating power loss / node crashes at multiple different points during the init sequence, to make sure that indexes or chainstate do not get corrupted - this has been a source of bugs in the past.

If you do fewer of them, the test will obviously run faster, but you also reduce the scope of the test.

hodlinator replied:
Thanks! I was relying too much on the log lines from the test saying the node would "exit" and not looking deeper into why we were calling `sigterm_node()`.

Dropped that commit and changed the log line to say "terminate" instead of "exit".

mzumsande replied:
oh, the tests uses SIGTERM - so it doesn't simulate a power loss like I said above, but a user abort - but the general point still stands. (I confused it with similar tests I sometimes run but never PR'ed which use SIGKILL).

hodlinator replied:
As this thread prompted me to understand the beginning of the test better, I realized I could take a stab at making it work on Windows, and found a way: #32021.

hodlinator committed:
qa: Add missing coverage of corrupt indexes

fjahr commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -70,7 +70,7 @@ def set_test_params(self):
         self.extra_args = [
             ["-txindex"],
-            ["-txindex"],
+            [],
```
Did you check if this has an performance impact and maybe that's why they were added here?

hodlinator replied:
At 081ae231ff542c254924a1770c9db666aff9f880^:
```shell
₿ hyperfine -w 1 "build/test/functional/rpc_rawtransaction.py --legacy-wallet"
Benchmark 1: build/test/functional/rpc_rawtransaction.py --legacy-wallet
  Time (mean ± σ):      4.776 s ±  0.025 s    [User: 1.420 s, System: 0.267 s]
  Range (min … max):    4.717 s …  4.811 s    10 runs
```
At 081ae231ff542c254924a1770c9db666aff9f880:
```shell
₿ hyperfine -w 1 "build/test/functional/rpc_rawtransaction.py --legacy-wallet"
Benchmark 1: build/test/functional/rpc_rawtransaction.py --legacy-wallet
  Time (mean ± σ):      4.761 s ±  0.029 s    [User: 1.387 s, System: 0.278 s]
  Range (min … max):    4.713 s …  4.791 s    10 runs
```
Conclusion: Negligible improvement.

More discussion in [other thread](https://github.com/bitcoin/bitcoin/pull/32010#discussion_r1985243645).

fjahr commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -117,7 +117,6 @@ def set_test_params(self):
         self.extra_args = [
             [
                 "-addresstype=bech32",
-                "-txindex",
```
Same as above, may check for performance impact on this test.

hodlinator replied:
At 081ae231ff542c254924a1770c9db666aff9f880^:
```shell
₿ hyperfine -w 1 "build/test/functional/wallet_avoid_mixing_output_types.py --descriptors"
Benchmark 1: build/test/functional/wallet_avoid_mixing_output_types.py --descriptors
  Time (mean ± σ):      1.581 s ±  0.588 s    [User: 0.610 s, System: 0.113 s]
  Range (min … max):    0.927 s …  2.288 s    10 runs
```

At 081ae231ff542c254924a1770c9db666aff9f880:
```shell
₿ hyperfine -w 1 "build/test/functional/wallet_avoid_mixing_output_types.py --descriptors"
Benchmark 1: build/test/functional/wallet_avoid_mixing_output_types.py --descriptors
  Time (mean ± σ):      1.640 s ±  0.575 s    [User: 0.602 s, System: 0.117 s]
  Range (min … max):    0.879 s …  2.288 s    10 runs
```
Conclusion: A slight increase in average run-time, but a faster min-run.

I prefer turning off functionality that isn't required for tests, eliminating them from this category of race conditions completely.

fjahr commented on test/functional/feature_init.py:
```diff
@@ -102,12 +102,21 @@ def check_clean_start():
             'blocks/index/*.ldb': 'Error opening block database.',
             'chainstate/*.ldb': 'Error opening coins database.',
             'blocks/blk*.dat': 'Error loading block database.',
+            # Not detected: 'indexes/blockfilter/basic/*.dat', 'indexes/blockfilter/basic/db/*.*', 'indexes/coinstats/db/*.*'
```
Can this be a bit more clear what this means? I guess "Removing these files does not result in a startup error"? Would be great if you could spell that out because I needed a moment to understand. You could also put the commented out line below and this one together and above the variable declaration. I am not sure if having them in line and split up helps with understanding. Same below.

hodlinator replied:
Done in latest push.

fjahr reviewed (commented):
Concept ACK

fjahr commented on test/functional/test_framework/util.py:
```diff
@@ -592,3 +592,10 @@ def find_vout_for_address(node, txid, addr):
     raise RuntimeError("Vout not found for address: txid=%s, addr=%s" % (txid, addr))
+
+
+def sync_txindex(test_framework, node):
```
nit: I would have put this into `test_framework/test_framework.py` so it can live among its peers `sync_blocks` et al and we would save the import from util. But fine to keep as is.

fjahr replied:
Right, I'm not passionate about the naming here. `sync_blocks` is syncing between two blockchains, `sync_txindex` is syncing between a blockchain and an index, so it's not wrong IMO. But `wait_for_txindex()` sounds fine, too. So up to you what you want to choose and if you want to change this at all.

fjahr reviewed (commented):
Code review ACK 12a1d5d813640addc8186e7fc882777f5a078164

hodlinator committed:
qa: Wait for txindex to avoid race condition

Can be verified to be necessary through adding std::this_thread::sleep_for(0.5s) at the beginning of TxIndex::CustomAppend.

fjahr commented:
re-ACK 3301d2cbe8c3b76c97285d75fa59637cb6952d0b

Only change was to make the helper internals a bit nicer.

mzumsande reviewed (commented):
Code Review ACK 3301d2cbe8c3b76c97285d75fa59637cb6952d0b

Prabhat1308 commented:
Are the files mentioned in `files_to_delete` and `files_to_perturb` just for informational purposes that they don't trigger an error or you want to indicate that these files should trigger an error but currently they don't.

hodlinator commented:
> Are the files mentioned in `files_to_delete` and `files_to_perturb` comments just for informational purposes that they don't trigger an error or you want to indicate that these files should trigger an error but currently they don't.

Yes, some of them at least probably should trigger an error, but I'm not knowledgeable of those aspects of LevelDB. It would probably be good to dig deeper into that, but I wanted to keep the focus of this PR on the tx index race conditions.

The reason I touched *feature_init.py* was because I was trying to eliminate unnecessary `-txindex` usage, eliminating the possibility for these race conditions - but it felt bad to just strip it out there, so I preferred to at least add file patterns in the test for what could be checked without changing the C++ code.

furszy reviewed (commented):
Code review ACK 3301d2cbe8c3b76c97285d75fa59637cb6952d0b

Prabhat1308 reviewed (commented):
Concept ACK [`3301d2c`](https://github.com/bitcoin/bitcoin/pull/32010/commits/3301d2cbe8c3b76c97285d75fa59637cb6952d0b)
I could not recreate a race condition but I have 
- Tested that the commented files dont trigger a startup error. 
- Ran all the tests to check that all pass .

hodlinator commented:
Thanks for testing!

> I could not recreate a race condition but I have

Did you try the recommended approach in the commit message of 3301d2cbe8c3b76c97285d75fa59637cb6952d0b? It is a bit terse.

* Make sure you don't have the added `sync_txindex` fix on the Python side.
* Add the `sleep_for()` in the recommended place in the C++ side, (build).
* You should be able to get the individual tests to fail (simplest is to run the test_runner.py suite so the `--args` get sent in correctly).

Have a feeling I needed slightly different sleep-duration on Windows/Linux due to differences in native/Python execution speed, so maybe it would help to insert slightly longer durations than recommended on Mac to reproduce the issue?

maflcko reviewed (commented):
lgtm, mod nit

hodlinator commented on test/functional/test_framework/util.py:
```diff
@@ -592,3 +592,10 @@ def find_vout_for_address(node, txid, addr):
     raise RuntimeError("Vout not found for address: txid=%s, addr=%s" % (txid, addr))
+
+
+def sync_txindex(test_framework, node):
```
It operates on the node, but I don't want test_node.py to contain the kitchen sink for all operations on a node. Some prior discussion about placement here: https://github.com/bitcoin/bitcoin/pull/32010#discussion_r1986065651
